### PR TITLE
Better Docker Auth lookup

### PR DESCRIPTION
**Pull request by dadgar**
This PR brings Nomad's Docker Auth lookup in sync with Docker's own implementation.

Fixes https://github.com/hashicorp/nomad/issues/2003
Fixes https://github.com/hashicorp/nomad/issues/1662

### CODE REVIEW

File: client/driver/docker.go [hunk 1]
```diff
@@ -1351,3 +1337,40 @@ func calculatePercent(newSample, oldSample, newTotal, oldTotal uint64, cores int
 
 	return (float64(numerator) / float64(denom)) * float64(cores) * 100.0
 }
+
+// authOptionsFrom takes the Docker auth config file and the repo being pulled
+// and returns an AuthConfiguration or an error if the file/repo could not be
+// parsed or looked up.
+func authOptionFrom(file, repo string) (*docker.AuthConfiguration, error) {
```
Comment (diptanu) [inline client/driver/docker.go]: I would like to see a test for this. Hard to follow what this method expects to return the right AuthConfiguration.

File: client/driver/docker.go [hunk 2]
```diff
@@ -1351,3 +1337,40 @@ func calculatePercent(newSample, oldSample, newTotal, oldTotal uint64, cores int
...
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse named repo %q: %v", err)
+	}
+
+	repoInfo, err := registry.ParseRepositoryInfo(name)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse repository: %v", err)
+	}
+
+	f, err := os.Open(file)
```
Comment (diptanu) [inline client/driver/docker.go]: Please close the file once you are done using it.

**Comment by thatsk** [comment]
How we can pass that ?

**Comment by thatsk** [comment]
if we have nomad cluster

**Comment by dadgar** [comment]
@thatsk https://www.nomadproject.io/docs/drivers/docker.html#docker-auth-config

**Comment by thatsk** [comment]
If i want to add in client.hcl or server.hcl

**Comment by dadgar** [comment]
You add to client.hcl:

```
client {
  options {
    "docker.auth.config" = "/path/to/config.json"
  }
}
```

**Comment by thatsk** [comment]
thanks dadgar
is this fix in version of nomad yet.?

**Comment by dadgar** [comment]
Yes 0.5.3 or greater should have this. Later versions of Nomad have fixed small bugs relating to this so would suggest running 0.5.6.

**Comment by parismorali** [comment]
Hi,

According to nomad documentation I should be able to do this

`config {
    image = "hello/world:latest"

    auth {
      username = "testuser"
      password = "testpassword"
      server_address = "my.docker.registry.com"
    }
  }`

and get a docker image from my login protected private docker registry.
However allocation log is

> Failed to pull `hello/world:latest`: API error (401): {"message":"Get https://registry-1.docker.io/v2/

It looks like nomad ignores the server_address I've speicified.

Am I doing something wrong?

Nomad version: 0.5.4

Thanks for your help
Paris

**Comment by dadgar** [comment]
Hey we are passing through that setting to the docker api client, which passes through to docker. Looks like they aren't honoring that and instead just parse the image name to detect the registry. I believe an example was given on the mailing list.

**Comment by parismorali** [comment]
Thanks, I've used the mailing list advice and this works.

**Comment by noxecane** [comment]
Just for anyone coming here to solve this issue
```hcl
config {
    image = "https://my.docker.registry.com/hello/world:latest"
    auth {
      username = "testuser"
      password = "testpassword"
    }
}
```

`server_address` I think is ignored
